src/adaptation/precond.jl

```diff
@@ -0,0 +1,152 @@
+mutable struct WelfordCovar{T<:AbstractFloat} <: CovarEstimator{T}
+    n :: Int
+    μ :: Vector{T}
+    M :: Matrix{T}
+end
+
+function reset!(wc::WelfordCovar{T}) where {T<:AbstractFloat}
+    wc.n = 0
+    wc.μ .= zero(T)
+    wc.M .= zero(T)
+end
+
+function add_sample!(wc::WelfordCovar, s::AbstractVector)
+    wc.n += 1
+    δ = s .- wc.μ
+    wc.μ .+= δ ./ wc.n
+    wc.M .+= (s .- wc.μ) * δ'
+end
+# Ref: https://github.com/stan-dev/stan/blob/develop/src/stan/mcmc/covar_adaptation.hpp
+function get_covar(wc::WelfordCovar{T})::Matrix{T} where {T<:AbstractFloat}
+    n, M = wc.n, wc.M
+    @assert n >= 2 "Cannot get variance with only one sample"
+    return (n / ((n + 5) * (n - 1))) .* M + 1e-3 * (5 / (n + 5)) * LinearAlgebra.I
+end
+
+################
+### Adapters ###
+################
+# TODO: integrate metric into preconditioner
+abstract type AbstractPreConditioner <: AbstractAdapter end
+struct UnitPreConditioner <: AbstractPreConditioner end
+
+string(::UnitPreConditioner) = "I"
+adapt!(::UnitPreConditioner, ::AbstractVector{<:Real}, ::AbstractFloat, is_update::Bool=true) = nothing
+reset!(::UnitPreConditioner) = nothing
+getM⁻¹(dpc::UnitPreConditioner) = nothing
+
+mutable struct DiagPreConditioner{T<:Real,AT<:AbstractVector{T}} <: AbstractPreConditioner
+    n_min   :: Int
+    ve  :: VarEstimator{T}
+    var :: AT
+end
+
+function DiagPreConditioner(d::Int, n_min::Int=10)
+    ve = WelfordVar(0, zeros(d), zeros(d))
+    return DiagPreConditioner(n_min, ve, Vector(ones(d)))
+end
+
+function string(dpc::DiagPreConditioner)
+    return string(dpc.var)
+end
+
+function adapt!(dpc::DiagPreConditioner, θ::AbstractVector{<:Real}, α::AbstractFloat, is_update::Bool=true)
+    add_sample!(dpc.ve, θ)
+    if dpc.ve.n >= dpc.n_min && is_update
+        dpc.var .= get_var(dpc.ve)
+    end
+end
+
+reset!(dpc::DiagPreConditioner) = reset!(dpc.ve)
+
+function getM⁻¹(dpc::DiagPreConditioner)
+    return dpc.var
+end
+
+mutable struct DensePreConditioner{T<:AbstractFloat} <: AbstractPreConditioner
+    n_min :: Int
+    ce    :: CovarEstimator{T}
+    covar :: Matrix{T}
+end
+
+function DensePreConditioner(d::Integer, n_min::Int=10)
+    ce = WelfordCovar(0, zeros(d), zeros(d,d))
+    # TODO: take use of the line below when we have an interface to set which pre-conditioner to use
+    # ce = NaiveCovar(0, Vector{Vector{Float64}}())
+    return DensePreConditioner(n_min, ce, LinearAlgebra.diagm(0 => ones(d)))
+end
+
+function string(dpc::DensePreConditioner)
+    return string(LinearAlgebra.diag(dpc.covar))
+end
+
+function adapt!(dpc::DensePreConditioner, θ::AbstractVector{<:AbstractFloat}, α::AbstractFloat, is_update::Bool=true)
+    add_sample!(dpc.ce, θ)
+    if dpc.ce.n >= dpc.n_min && is_update
+        dpc.covar .= get_covar(dpc.ce)
+    end
+end
+
+reset!(dpc::DensePreConditioner) = reset!(dpc.ce)
+
+function getM⁻¹(dpc::DensePreConditioner)
+    return dpc.covar
+end
```

Comment:
This file only contain type definition changes and interface changes towards the original codes from Turing.jl.